File: src/termer/OutputterTest.hs

module OutputterTest (test)
  where

import Control.Monad.State

import Outputter
import Screen
import Test

--  Run a sequence of commands on a given sized terminal and return the result.
runexp :: Integer -> Integer -> State Screen () -> Screen
runexp cls lns x = snd $ runState x (screen cls lns)


type TS = (String, Screen)

getf :: State TS Char
getf = do
    s <- gets fst
    if null s
      then return '.'
      else (modify $ \(_, scr) -> (tail s, scr)) >> return (head s)

updatef :: (Screen -> Screen) -> State TS () 
updatef f = modify $ \(str, scr) -> (str, f scr)

-- Run an outputter on the given characters and sized terminal and return
-- resulting screen.
runwnt :: Integer -> Integer -> String -> Screen
runwnt cls lns chrs = snd . snd $ runState (outputter (Just '.') getf updatef) (chrs, screen cls lns)


-- Test an outputter test by passing the chars to an outputter, and running the
-- commands manually, and checking to make sure the resulting screens are the
-- same.
run :: Integer -> Integer -> String -> State Screen () -> Test
run cls lns chrs x = runexp cls lns x ~=? runwnt cls lns chrs

test = "OutputterTest" ~: [
    "simple" ~: (run 6 4 "a" $ do
        modify $ put_char 'a'
        ),

    "multi" ~: (run 6 4 "abc" $ do
        modify $ put_char 'a'
        modify $ put_char 'b'
        modify $ put_char 'c'
        ),

    "newline" ~: (run 6 4 "ab\n\rc" $ do
        modify $ put_char 'a'
        modify $ put_char 'b'
        modify $ newline
        modify $ put_char 'c'
        ),

    "backspace" ~: (run 6 4 "ab\BSc" $ do
        modify $ put_char 'a'
        modify $ put_char 'b'
        modify $ cursor_left
        modify $ put_char 'c'
        ),

    "ESC[m" ~: (run 6 4 "ab\ESC[mc" $ do
        modify $ put_char 'a'
        modify $ put_char 'b'
        modify $ exit_attribute_mode
        modify $ put_char 'c'
        ),

    "ESC[0m" ~: (run 6 4 "ab\ESC[0mc" $ do
        modify $ put_char 'a'
        modify $ put_char 'b'
        modify $ exit_attribute_mode
        modify $ put_char 'c'
        ),

    "sgr: ESC[0;10m" ~: (run 6 4 "a\ESC[1mb\ESC[0;10mc" $ do
        modify $ put_char 'a'
        modify $ enter_bold_mode
        modify $ put_char 'b'
        modify $ exit_attribute_mode
        modify $ put_char 'c'
        ),

    "boldoff: ESC[22m" ~: (run 6 4 "a\ESC[1mb\ESC[22mc" $ do
        modify $ put_char 'a'
        modify $ enter_bold_mode
        modify $ put_char 'b'
        modify $ exit_bold_mode
        modify $ put_char 'c'
        ),

    "cup: ESC[4;3H" ~: (run 6 4 "\ESC[4;3H" $ do
        modify $ cursor_address (Position 3 4)
        ),

    "cuf: ESC[4C" ~: (run 6 4 "\ESC[4C" $ do
        modify $ parm_right_cursor 4
        )

    ]

File: src/termer/outputtertest.c

// Copyright (C) 2011 Richard Uhler <ruhler@member.fsf.org>
//
// This file is part of Focus.
//
// Focus is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Focus is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Focus.  If not, see <http://www.gnu.org/licenses/>.

#include <stdio.h>

#include "outputter.h"

const char* runstringbuf = NULL;

char runstringgetf()
{
    char c = *runstringbuf;
    runstringbuf++;
    return c;
}

// update the given screen by running outputter on the given string.
void runstring(SCREEN_Screen* scr, const char* str)
{
    runstringbuf = str;
    outputter(scr, '\0', runstringgetf);
}

void update(bool* result, bool x)
{
    *result = *result && x;
}

bool tscreq(const char* name, SCREEN_Screen* a, SCREEN_Screen* b)
{
    if (!eq_screen(a, b)) {
        printf("%s:\n", name);

        if (!eq_position(&a->cursor, &b->cursor)) {
            printf("\tecursor = %i,%i\n", a->cursor.column, a->cursor.line);
            printf("\tgcursor = %i,%i\n", b->cursor.column, b->cursor.line);
        }

        if (!eq_attributes(&a->sattrs, &b->sattrs)) {
            printf("\teattr = fg %i, bg %i, b %i, r %i\n",
                    a->sattrs.fgcolor, a->sattrs.bgcolor,
                    a->sattrs.style.bold, a->sattrs.style.reverse);
            printf("\tgattr = fg %i, bg %i, b %i, r %i\n",
                    b->sattrs.fgcolor, b->sattrs.bgcolor,
                    b->sattrs.style.bold, b->sattrs.style.reverse);
        }

        int x, y;
        for (y = 0; y < a->lines; y++) {
            for (x = 0; x < a->columns; x++) {
                SCREEN_Cell acell = cellat(a, mkpos(x, y));
                SCREEN_Cell bcell = cellat(b, mkpos(x, y));
                if (!eq_cell(&acell, &bcell)) {
                    printf("\tecell %i,%i = %c(%i), fg %i, bg %i, b %i, r %i\n",
                            x, y, acell.character, acell.character,
                            acell.cattrs.fgcolor, acell.cattrs.bgcolor,
                            acell.cattrs.style.bold, acell.cattrs.style.reverse);
                    printf("\tgcell %i,%i = %c(%i) fg %i bg %i, b %i, r %i\n",
                            x, y, bcell.character, bcell.character,
                            bcell.cattrs.fgcolor, bcell.cattrs.bgcolor,
                            bcell.cattrs.style.bold, bcell.cattrs.style.reverse);
                }
            }
        }

        return false;
    }
    return true;
}

void test(bool* result)
{
    SCREEN_Screen wnt;
    SCREEN_Screen got;

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "a");
    put_char(&wnt, 'a');
    update(result, tscreq("simple", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "abc");
    put_char(&wnt, 'a');
    put_char(&wnt, 'b');
    put_char(&wnt, 'c');
    update(result, tscreq("multi", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "ab\n\rc");
    put_char(&wnt, 'a');
    put_char(&wnt, 'b');
    newline(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("newline", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "ab\bc");
    put_char(&wnt, 'a');
    put_char(&wnt, 'b');
    cursor_left(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("backspace", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "ab\e[mc");
    put_char(&wnt, 'a');
    put_char(&wnt, 'b');
    exit_attribute_mode(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("ESC[m", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "ab\e[0mc");
    put_char(&wnt, 'a');
    put_char(&wnt, 'b');
    exit_attribute_mode(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("ESC[0m", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "a\e[1mb\e[0;10mc");
    put_char(&wnt, 'a');
    enter_bold_mode(&wnt);
    put_char(&wnt, 'b');
    exit_attribute_mode(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("sgr: ESC[0;10m", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "a\e[1mb\e[22mc");
    put_char(&wnt, 'a');
    enter_bold_mode(&wnt);
    put_char(&wnt, 'b');
    exit_bold_mode(&wnt);
    put_char(&wnt, 'c');
    update(result, tscreq("boldoff: ESC[22m", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "\e[4;3H");
    cursor_address(&wnt, mkpos(2,3));
    update(result, tscreq("cup: ESC[4;3H", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "\e[4C");
    parm_right_cursor(&wnt, 4);
    update(result, tscreq("cuf: ESC[4C", &wnt, &got));

    wnt = screen(24, 4);
    got = screen(24, 4);
    runstring(&got, "a\tb");
    put_char(&wnt, 'a');
    tab(&wnt);
    put_char(&wnt, 'b');
    update(result, tscreq("tab", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "a\e[1;34mb");
    put_char(&wnt, 'a');
    enter_bold_mode(&wnt);
    set_foreground(&wnt, SCREEN_COLOR_BLUE);
    put_char(&wnt, 'b');
    update(result, tscreq("ESC[1;34m", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "\e[42ma\e[49mb");
    set_background(&wnt, SCREEN_COLOR_GREEN);
    put_char(&wnt, 'a');
    set_background(&wnt, SCREEN_COLOR_BLACK);
    put_char(&wnt, 'b');
    update(result, tscreq("default background color", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "\xE2\x80\x98m");
    put_char(&wnt, 0x2018);
    put_char(&wnt, 'm');
    update(result, tscreq("unicode", &wnt, &got));

    wnt = screen(6, 4);
    got = screen(6, 4);
    runstring(&got, "\e[3;3Ha\eM");
    cursor_address(&wnt, mkpos(2, 2));
    put_char(&wnt, 'a');
    scroll_reverse(&wnt);
    update(result, tscreq("ESC M", &wnt, &got));
}

int main()
{
    bool result = true;
    test(&result);
    if (result) {
        printf("all tests passed.\n");
        return 0;
    } else {
        printf("there were failing tests.\n");
        return 1;
    }
}
